File: WinPaletter/GUI/Elements/Input/NumericUpDown.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D
Imports System.Drawing.Text

Namespace UI.WP

    <Description("Themed NumericUpDown for WinPaletter UI")> Public Class NumericUpDown : Inherits Control

        Sub New()
            Font = New Font("Segoe UI", 9)
            Enabled = True
            DoubleBuffered = True
        End Sub

#Region "Variables"

        Private _Shown As Boolean = False
        Private SideRect As New Rectangle

        Dim State As MouseState = MouseState.None

        Enum MouseState
            None
            Over
            Down
        End Enum

#End Region

#Region "Properties"

        Public Property UpDownStep As Integer = 1

        Private _Value As Integer
        Public Property Value As Integer
            Get
                Return _Value
            End Get
            Set(value As Integer)
                Select Case value
                    Case Is > Max
                        value = Max
                        Invalidate()

                    Case Is < Min
                        value = Min
                        Invalidate()
                End Select
                _Value = value
                Invalidate()
                RaiseEvent ValueChanged(Me, EventArgs.Empty)
            End Set
        End Property

        Private _Max As Integer = 100
        Public Property Max As Integer
            Get
                Return _Max
            End Get
            Set(value As Integer)
                Select Case value
                    Case Is < _Value
                        _Value = value
                End Select
                _Max = value
                Invalidate()
            End Set
        End Property

        Private _Min As Integer
        Public Property Min As Integer
            Get
                Return _Min
            End Get
            Set(value As Integer)
                Select Case value
                    Case Is > _Value
                        _Value = value
                End Select
                _Min = value
                Invalidate()
            End Set
        End Property

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String = ""

#End Region

#Region "Animator"

        Dim alpha As Integer
        ReadOnly Factor As Integer = 20
        Dim WithEvents Timer As New Timer With {.Enabled = False, .Interval = 1}

        Private Sub Timer_Tick(sender As Object, e As EventArgs) Handles Timer.Tick
            If Not DesignMode Then

                If State = MouseState.Over Then
                    If alpha + Factor <= 255 Then
                        alpha += Factor
                    ElseIf alpha + Factor > 255 Then
                        alpha = 255
                        Timer.Enabled = False
                        Timer.Stop()
                    End If

                    If _Shown Then
                        Threading.Thread.Sleep(1)
                        Invalidate()
                    End If
                End If

                If Not State = MouseState.Over Then
                    If alpha - Factor >= 0 Then
                        alpha -= Factor
                    ElseIf alpha - Factor < 0 Then
                        alpha = 0
                        Timer.Enabled = False
                        Timer.Stop()
                    End If

                    If _Shown Then
                        Threading.Thread.Sleep(1)
                        Invalidate()
                    End If
                End If
            End If
        End Sub

#End Region

#Region "Events"

        Public Event ValueChanged(sender As Object, e As EventArgs)

        Protected Overrides Sub OnMouseEnter(e As EventArgs)
            MyBase.OnMouseEnter(e)
            State = MouseState.Over
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Protected Overrides Sub OnMouseLeave(e As EventArgs)
            MyBase.OnMouseLeave(e)
            State = MouseState.None
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Protected Overrides Sub OnMouseUp(e As MouseEventArgs)
            State = MouseState.Over
            _Shown = True
            Timer.Enabled = True
            Timer.Start()

            MyBase.OnMouseUp(e)

            If Enabled Then
                If SideRect.Contains(e.Location) And e.Y < 10 Then
                    Value += UpDownStep
                ElseIf SideRect.Contains(e.Location) And e.Y > 10 Then
                    Value -= UpDownStep
                End If
            End If
        End Sub

        Protected Overrides Sub OnResize(e As EventArgs)
            Me.Invalidate()
        End Sub

        Private Sub NumericUpDown_MouseDown(sender As Object, e As MouseEventArgs) Handles Me.MouseDown
            State = MouseState.Down
            _Shown = True

            If Enabled And SideRect.Contains(e.Location) Then
                Timer.Enabled = True
                Timer.Start()
            End If
        End Sub

        Private Sub NumericUpDown_HandleCreated(sender As Object, e As EventArgs) Handles Me.HandleCreated
            alpha = 0

            If Not DesignMode Then
                Try
                    AddHandler FindForm.Load, AddressOf Loaded
                    AddHandler FindForm.Shown, AddressOf Showed
                    AddHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                Catch
                End Try
            End If
        End Sub

        Private Sub NumericUpDown_HandleDestroyed(sender As Object, e As EventArgs) Handles Me.HandleDestroyed
            If Not DesignMode Then
                Try
                    RemoveHandler FindForm.Load, AddressOf Loaded
                    RemoveHandler FindForm.Shown, AddressOf Showed
                    RemoveHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                Catch
                End Try
            End If
        End Sub

        Sub Loaded(sender As Object, e As EventArgs)
            _Shown = False
        End Sub

        Sub Showed(sender As Object, e As EventArgs)
            _Shown = True
            Invalidate()
        End Sub

        Public Sub RefreshColorPalette(sender As Object, e As EventArgs)
            If _Shown Then
                Invalidate()
            End If
        End Sub

#End Region

        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Dim G As Graphics = e.Graphics
            G.SmoothingMode = SmoothingMode.AntiAlias
            G.TextRenderingHint = If(DesignMode, TextRenderingHint.ClearTypeGridFit, TextRenderingHint.SystemDefault)
            DoubleBuffered = True
            Dim RTL As Boolean = (RightToLeft = 1)

            '################################################################################# Customizer
            Dim OuterRect As New Rectangle(0, 0, Width - 1, Height - 1)
            Dim InnerRect As New Rectangle(1, 1, Width - 3, Height - 3)
            SideRect = New Rectangle(Width - 16, 0, 15, Height)

            If RTL Then
                OuterRect.X = Width - OuterRect.X - OuterRect.Width
                InnerRect.X = Width - InnerRect.X - InnerRect.Width
                SideRect.X = Width - SideRect.X - SideRect.Width
            End If

            '#################################################################################

            G.Clear(GetParentColor)

            Using br As New SolidBrush(Color.FromArgb(255 - alpha, My.Style.Colors.Back)) : G.FillRoundedRect(br, OuterRect) : End Using
            Using br As New SolidBrush(Color.FromArgb(alpha, My.Style.Colors.Back_Checked)) : G.FillRoundedRect(br, OuterRect) : End Using
            Using br As New SolidBrush(Color.FromArgb(alpha, My.Style.Colors.Border_Checked_Hover)) : G.FillRoundedRect(br, SideRect) : End Using

            Using P As New Pen(Color.FromArgb(255 - alpha, My.Style.Colors.Border)) : G.DrawRoundedRect_LikeW11(P, InnerRect) : End Using
            Using P As New Pen(Color.FromArgb(alpha, My.Style.Colors.Border_Checked_Hover)) : G.DrawRoundedRect_LikeW11(P, OuterRect) : End Using

            If Focused And State = MouseState.None Then
                Using P As New Pen(Color.FromArgb(255, My.Style.Colors.Border_Checked_Hover)) : G.DrawRoundedRect(P, InnerRect) : End Using
            End If

            Using TextColor As New SolidBrush(If(My.Style.DarkMode, Color.White, Color.Black))
                G.DrawString(CStr(Value), Font, TextColor, New Rectangle(0, 0, Width - 15, Height), ContentAlignment.MiddleCenter.ToStringFormat)
            End Using

            Using SignColor As New SolidBrush(My.Style.Colors.Back_Checked), SignFont As New Font("Marlett", 11)
                G.DrawString("t", SignFont, SignColor, New Point(SideRect.Left - 1, 0))
                G.DrawString("u", SignFont, SignColor, New Point(SideRect.Left - 1, Height - 16))
            End Using
        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/Input/NumericUpDown.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Windows.Forms;

namespace WinPaletter.UI.WP
{
    [Description("Themed NumericUpDown for WinPaletter UI")]
    public class NumericUpDown : Control
    {
        public NumericUpDown()
        {
            SetStyle(ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
            DoubleBuffered = true;
            BackColor = Color.Transparent;
        }

        #region Variables
        private bool CanAnimate => !DesignMode && Program.Style.Animations && this != null && Visible && Parent != null && Parent.Visible && FindForm() != null && FindForm().Visible;

        private Rectangle SideRect = new();

        private MouseState State = MouseState.None;

        public enum MouseState
        {
            None,
            Over,
            Down
        }

        #endregion

        #region Properties

        public int UpDownStep { get; set; } = 1;

        private int _Value;
        public int Value
        {
            get => _Value;
            set
            {
                if (value != _Value)
                {
                    if (value > Maximum) value = Maximum;
                    if (value < Minimum) value = Minimum;
                    _Value = value;
                    Invalidate();
                    ValueChanged?.Invoke(this, EventArgs.Empty);
                }
            }
        }


        private int _Max = 100;
        public int Maximum
        {
            get => _Max;
            set
            {
                if (value != _Max)
                {
                    _Max = value;
                    if (value < _Value) Value = value;
                    Invalidate();
                }
            }
        }


        private int _Min;
        public int Minimum
        {
            get => _Min;
            set
            {
                if (value != _Min)
                {
                    _Min = value;
                    if (value > _Value) Value = value;
                    Invalidate();
                }
            }
        }

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(System.ComponentModel.Design.MultilineStringEditor), typeof(System.Drawing.Design.UITypeEditor))]
        [Bindable(true)]
        public override string Text { get; set; } = string.Empty;

        #endregion

        #region Events/Overrides

        public event ValueChangedEventHandler ValueChanged;

        public delegate void ValueChangedEventHandler(object sender, EventArgs e);

        protected override void OnMouseWheel(MouseEventArgs e)
        {
            if (e.Delta < 0)
            {
                if (e.Delta <= -240)
                    Value = Math.Max(Minimum, Value - UpDownStep * 2);
                else
                    Value = Math.Max(Minimum, Value - UpDownStep);
            }
            else
            {
                if (e.Delta >= 240)
                    Value = Math.Min(Maximum, Value + UpDownStep * 2);
                else
                    Value = Math.Min(Maximum, Value + UpDownStep);
            }

            base.OnMouseWheel(e);
        }

        protected override void OnMouseEnter(EventArgs e)
        {
            State = MouseState.Over;

            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), 255).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 255; }

            base.OnMouseEnter(e);
        }

        protected override void OnMouseLeave(EventArgs e)
        {
            State = MouseState.None;

            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 0; }

            base.OnMouseLeave(e);
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            State = MouseState.Over;


            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), ContainsFocus ? 255 : 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = ContainsFocus ? 255 : 0; }

            if (Enabled)
            {
                if (SideRect.Contains(e.Location) & e.Y < 10)
                {
                    Value += UpDownStep;
                }
                else if (SideRect.Contains(e.Location) & e.Y > 10)
                {
                    Value -= UpDownStep;
                }
            }

            base.OnMouseUp(e);
        }

        protected override void OnResize(EventArgs e)
        {
            Invalidate();

            base.OnResize(e);
        }

        protected override void OnMouseDown(MouseEventArgs e)
        {
            State = MouseState.Down;

            if (Enabled & SideRect.Contains(e.Location))
            {
                if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
                else { alpha = 0; }
            }

            base.OnMouseDown(e);
        }

        int parentLevel = 0;
        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);

            parentLevel = this.Level();
        }


        #endregion

        #region Animator
        private int _alpha = 0;
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public int alpha
        {
            get => _alpha;
            set { _alpha = value; Invalidate(); }
        }
        #endregion

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            //Leave it empty to make control background transparent
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.AntiAlias;
            G.TextRenderingHint = DesignMode ? TextRenderingHint.ClearTypeGridFit : Program.Style.TextRenderingHint;
            bool RTL = (int)RightToLeft == 1;

            //Makes background drawn properly, and transparent
            InvokePaintBackground(this, e);

            Config.Scheme scheme = Enabled ? Program.Style.Schemes.Main : Program.Style.Schemes.Disabled;

            // ################################################################################# Customizer
            Rectangle OuterRect = new(0, 0, Width - 1, Height - 1);
            Rectangle InnerRect = new(1, 1, Width - 3, Height - 3);
            SideRect = new(Width - 18, 0, 17, Height);

            if (RTL)
            {
                OuterRect.X = Width - OuterRect.X - OuterRect.Width;
                InnerRect.X = Width - InnerRect.X - InnerRect.Width;
                SideRect.X = Width - SideRect.X - SideRect.Width;
            }
            // #################################################################################

            using (SolidBrush br = new(scheme.Colors.Back(parentLevel)))
            {
                G.FillRoundedRect(br, InnerRect);
            }

            using (SolidBrush br = new(Color.FromArgb(alpha, scheme.Colors.Back_Checked))) { G.FillRoundedRect(br, OuterRect); }

            using (SolidBrush br = new(Color.FromArgb(alpha, scheme.Colors.Line_Checked_Hover))) { G.FillRoundedRect(br, SideRect); }

            using (Pen P = new(Color.FromArgb(255 - alpha, scheme.Colors.Line(parentLevel)))) { G.DrawRoundedRectBeveled(P, InnerRect); }

            using (Pen P = new(Color.FromArgb(alpha, scheme.Colors.Line_Checked_Hover))) { G.DrawRoundedRectBeveled(P, OuterRect); }

            using (SolidBrush SignBrush = new(Color.FromArgb(255 - alpha, scheme.Colors.Line_Checked_Hover)))
            {
                using (Font SignFont = new("Marlett", 11f))
                {
                    G.DrawString("t", SignFont, scheme.Brushes.Back_Checked, new Point(SideRect.Left, 0));
                    G.DrawString("u", SignFont, scheme.Brushes.Back_Checked, new Point(SideRect.Left, Height - 16));

                    G.DrawString("t", SignFont, SignBrush, new Point(SideRect.Left, 0));
                    G.DrawString("u", SignFont, SignBrush, new Point(SideRect.Left, Height - 16));
                }
            }

            using (StringFormat sf = ContentAlignment.MiddleCenter.ToStringFormat())
            {
                using (SolidBrush TextColor = new(ForeColor))
                {
                    G.DrawString($"{Value}", Fonts.ConsoleMedium, TextColor, new Rectangle(0, 1, Width - SideRect.Width, Height), sf);
                }
            }

            base.OnPaint(e);


        }
    }
}